File: pc_nodes_examples/cpp/src/mpu6050_sub.cpp
#include <zenoh_ros/ZenohRosPC.h>
#include <zenoh_ros/sensor_msgs/z_Imu.h>
#include <iomanip>
#include <sstream>

class MPU6050SubscriberNode : public ZenohNode {
public:
    MPU6050SubscriberNode() : ZenohNode("mpu6050_subscriber") {
        ZLOG_INFO(this->get_logger(), "Node has been started");

        // Create subscription to topic "robot/mpu6050" with standard ROS 2 z_Imu interface
        sub_ = this->z_create_subscription<z_Imu>(
            "robot/mpu6050",
            [this](const z_Imu& msg) -> void {
                this->listener_callback(msg);
            },
            10
        );
    }

private:
    ZenohSubscription<z_Imu>* sub_ = nullptr;

    void listener_callback(const z_Imu& msg) {
        std::ostringstream oss;
        oss << std::fixed << std::setprecision(2)
            << "[IMU RECV] [stamp: " << msg.header.stamp.sec << "." << msg.header.stamp.nanosec
            << "] Accel: ("
            << std::setw(6) << msg.linear_acceleration.x << ", "
            << std::setw(6) << msg.linear_acceleration.y << ", "
            << std::setw(6) << msg.linear_acceleration.z << ") m/s² | Gyro: ("
            << std::setw(6) << msg.angular_velocity.x << ", "
            << std::setw(6) << msg.angular_velocity.y << ", "
            << std::setw(6) << msg.angular_velocity.z << ") rad/s";
        ZLOG_INFO(this->get_logger(), "%s", oss.str().c_str());
    }
};

int main(int argc, char** argv) {
    (void)argc;
    (void)argv;

    // Connect to ESP32-S3 SoftAP via Wi-Fi
    ZenohConfig config;
    config.communication_mode = ZenohCommunicationMode::ZENOH_COMM_WIFI;
    config.host = "192.168.4.1";
    config.port = 7447;

    if (!ZenohNode::init(config)) {
        return -1;
    }

    {
        MPU6050SubscriberNode node;
        node.z_spin();
    }

    return 0;
}

File: pc_nodes_examples/python/mpu6050_sub.py
#!/usr/bin/env python3
import sys
import os

# Ensure the shared zenoh_ros package can be imported
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../../shared_libraries/python')))

from zenoh_ros import ZenohNode, ZenohConfig
from zenoh_ros.sensor_msgs import z_Imu

class MPU6050SubscriberNode(ZenohNode):
    def __init__(self) -> None:
        super().__init__("mpu6050_subscriber")
        self.get_logger().info("Node has been started")

        # Create subscription to topic "robot/mpu6050" with standard ROS 2 z_Imu type
        self.sub = self.z_create_subscription(
            z_Imu,
            "robot/mpu6050",
            self.listener_callback,
            10
        )

    def listener_callback(self, msg: z_Imu) -> None:
        self.get_logger().info(
            f"[IMU RECV] [stamp: {msg.header.stamp.sec}.{msg.header.stamp.nanosec:09d}] "
            f"Accel: ({msg.linear_acceleration.x:6.2f}, {msg.linear_acceleration.y:6.2f}, {msg.linear_acceleration.z:6.2f}) m/s² | "
            f"Gyro: ({msg.angular_velocity.x:6.2f}, {msg.angular_velocity.y:6.2f}, {msg.angular_velocity.z:6.2f}) rad/s"
        )


def main() -> None:
    # Connect to ESP32-S3 SoftAP IP (192.168.4.1)
    cfg = ZenohConfig(communication_mode=ZenohConfig.ZENOH_COMM_WIFI, host="192.168.4.1", port=7447)
    if not ZenohNode.init(cfg):
        return
    node_instance = None
    try:
        node_instance = MPU6050SubscriberNode()
        node_instance.z_spin()

    except KeyboardInterrupt:
        if node_instance:
            node_instance.get_logger().info("Shutdown requested via KeyboardInterrupt.")

    except Exception as e:
        if node_instance:
            node_instance.get_logger().error(f"Critical Error: {e}")

    finally:
        ZenohNode.shutdown()



if __name__ == "__main__":
    main()
